third_party/gtp_ovs/ovs-gtp-patches/2.15/0023-QFI-Support-in-OVS.patch

```diff
@@ -0,0 +1,1034 @@
+             error = parse_bundle_load(value, pp);
++        } else if (!strcasecmp(key, "qfi")) {
++            error = parse_set_tunnel_qfi(value, pp);
+         } else if (!strcasecmp(key, "drop")) {
+             drop = true;
+         } else if (!strcasecmp(key, "apply_actions")) {
+diff --git a/lib/packets.h b/lib/packets.h
+index 6edf85f..23e9a40 100644
+--- a/lib/packets.h
++++ b/lib/packets.h
+@@ -1463,6 +1463,14 @@ struct gtpuhdr_opt {
+ };
+ BUILD_ASSERT_DECL(sizeof(struct gtpuhdr_opt) == 4);
+ 
++struct gtpu_ext_hdr {
++    uint8_t len;
++    uint8_t pdu_type;
++    uint8_t qfi;
++    uint8_t next_type;
++};
++BUILD_ASSERT_DECL(sizeof(struct gtpu_ext_hdr) == 4);
++
+ /* VXLAN protocol header */
+ struct vxlanhdr {
+     union {
+diff --git a/lib/tc.h b/lib/tc.h
+index 281231c..8c44262 100644
+--- a/lib/tc.h
++++ b/lib/tc.h
+@@ -211,6 +211,7 @@ struct tc_action {
+             uint8_t tos;
+             uint8_t ttl;
+             uint8_t no_csum;
++            uint8_t qfi;
+             struct {
+                 ovs_be32 ipv4_src;
+                 ovs_be32 ipv4_dst;
+diff --git a/ofproto/tunnel.c b/ofproto/tunnel.c
+index 42a458b..695d948 100644
+--- a/ofproto/tunnel.c
++++ b/ofproto/tunnel.c
+@@ -45,10 +45,12 @@ struct tnl_match {
+     ovs_be64 in_key;
+     struct in6_addr ipv6_src;
+     struct in6_addr ipv6_dst;
++    uint8_t qfi;
+     odp_port_t odp_port;
+     bool in_key_flow;
+     bool ip_src_flow;
+     bool ip_dst_flow;
++    bool qfi_flow;
+     enum netdev_pt_mode pt_mode;
+ };
+ 
+@@ -167,9 +169,11 @@ tnl_port_add__(const struct ofport_dpif *ofport, const struct netdev *netdev,
+     tnl_port->match.in_key = cfg->in_key;
+     tnl_port->match.ipv6_src = cfg->ipv6_src;
+     tnl_port->match.ipv6_dst = cfg->ipv6_dst;
++    tnl_port->match.qfi = cfg->qfi;
+     tnl_port->match.ip_src_flow = cfg->ip_src_flow;
+     tnl_port->match.ip_dst_flow = cfg->ip_dst_flow;
+     tnl_port->match.in_key_flow = cfg->in_key_flow;
++    tnl_port->match.qfi_flow = cfg->qfi_present;
+     tnl_port->match.odp_port = odp_port;
+     tnl_port->match.pt_mode = netdev_get_pt_mode(netdev);
+ 
+@@ -381,7 +385,7 @@ tnl_wc_init(struct flow *flow, struct flow_wildcards *wc)
+          * wildcarded, not to unwildcard them here. */
+         wc->masks.tunnel.tp_src = 0;
+         wc->masks.tunnel.tp_dst = 0;
+-
++        wc->masks.tunnel.qfi = 0;
+         if (is_ip_any(flow)
+             && IP_ECN_is_ce(flow->tunnel.ip_tos)) {
+             wc->masks.nw_tos |= IP_ECN_MASK;
+@@ -436,7 +440,9 @@ tnl_port_send(const struct ofport_dpif *ofport, struct flow *flow,
+     if (!cfg->out_key_flow) {
+         flow->tunnel.tun_id = cfg->out_key;
+     }
+-
++    if (!cfg->qfi_present) {
++        flow->tunnel.qfi = cfg->qfi;
++    }
+     if (!cfg->out_key_flow && !cfg->out_key_present) {
+         /* since OAM is never set via OVSDB, do not touch that bit. */
+         flow->tunnel.flags &= FLOW_TNL_F_OAM;
```

Comment:
Lets add gtp unit tests odp.at to validate the parsing of qfi field.

Reply:
Added UT for qfi matching and action criteria.